### PR TITLE
[FIXED JENKINS-36144] Borrow the SCMTrigger's queue

### PR DESCRIPTION
**Pull request by stephenc**
https://issues.jenkins-ci.org/browse/JENKINS-36144

@reviewbybees @jenkinsci/code-reviewers

---
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/126)

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java [hunk 1]
```diff
@@ -235,6 +241,47 @@ public void writeLogTo(XMLOutput out) throws IOException {
...
+         */
+        /*package*/
+        synchronized void checkThreadPoolSize() {
+            if (scmTrigger != null) {
+                int count = scmTrigger.getPollingThreadCount();
+                if (maximumThreads != count) {
+                    maximumThreads = count;
+                    try {
+                        Field getQueue = SCMTrigger.DescriptorImpl.class.getDeclaredField("queue");
+                        getQueue.setAccessible(true);
```
Comment (KostyaSha) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: hack?

Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: of course... but after a more careful analysis I decided against that and rather just mirror the pool size... which sadly means you have two pools... but this does point that perhaps we should have an explicit pool for polling that plugins can use rather than the current status where each push notifier has to roll their own... with hillarious consequences

**Comment by jtnord** [comment]
:bee: @stephenc I assume you are going to file a issue for the generic issue

> but after a more careful analysis I decided against that and rather just mirror the pool size... which sadly means you have two pools... but this does point that perhaps we should have an explicit pool for polling that plugins can use rather than the current status where each push notifier has to roll their own... with hillarious consequences

**Comment by rsandell** [comment]
:bee: A future further improvement could perhaps also be to not schedule the polling if the same repo and branch is already in the queue.

**Comment by KostyaSha** [comment]
Queue skipping polling for the same project AFAIK

**Comment by KostyaSha** [comment]
> A future further improvement could perhaps also be to not schedule the polling if the same repo and branch is already in the queue.

Could you describe how queue works now?

**Comment by stephenc** [comment]
https://issues.jenkins-ci.org/browse/JENKINS-36147 filed @jtnord

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java [hunk 2]
```diff
@@ -333,6 +362,10 @@ public static boolean allowsHookUrlOverride() {
             return ALLOW_HOOKURL_OVERRIDE;
         }
 
+        private static ThreadFactory threadFactory() {
```
Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: it should be method that returns executors - `threadPoolWithCapacity(int)`

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java [hunk 3]
```diff
@@ -235,6 +239,31 @@ public void writeLogTo(XMLOutput out) throws IOException {
...
+        }
+
+        /**
+         * Update the {@link java.util.concurrent.ExecutorService} instance.
+         */
+        /*package*/
+        synchronized void checkThreadPoolSize() {
+            if (scmTrigger != null) {
+                int count = scmTrigger.getPollingThreadCount();
+                if (maximumThreads != count) {
```
Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: why just not `count >= 0`?

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java [hunk 4]
```diff
@@ -235,6 +239,31 @@ public void writeLogTo(XMLOutput out) throws IOException {
...
+
+        public DescriptorImpl() {
+            checkThreadPoolSize();
+        }
+
+        /**
+         * Update the {@link java.util.concurrent.ExecutorService} instance.
+         */
+        /*package*/
+        synchronized void checkThreadPoolSize() {
```
Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: should be named `updateThreadPoolSize` as of it not only checks it

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java [hunk 5]
```diff
@@ -235,6 +239,31 @@ public void writeLogTo(XMLOutput out) throws IOException {
...
+            if (scmTrigger != null) {
+                int count = scmTrigger.getPollingThreadCount();
+                if (maximumThreads != count) {
+                    queue.setExecutors(
+                            (count == 0
+                                    ? Executors.newCachedThreadPool(threadFactory())
+                                    : Executors.newFixedThreadPool(maximumThreads, threadFactory())));
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: Does this not cause a thread leak?
Say `ScmTrigger.pollingThreadCount` was set to `100`, this kicks in and creates a new fixed size thread pool, then you have a lot of notifcations and you use all 100 Threads...
then the admin sees this and sets `ScmTrigger.pollingThreadCount` to a more sane value of `10`, and this creates a new fixed size thread pool Executor...

But the original ExecutorService (with 100 threads) is not shutdown, and not does it wait for all the tasks to complete...

Basically you are waiting for GC to kick (and hoping that it does the right thing)...

**Comment by jtnord** [comment]
:bug: as I think the Executor service has the potential to leak threads (at least temporarily until some GC kicks in, the point of which is non-deterministic)

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java (hunk 5 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: When you call `queue.setExecutor(queue.getExecutor())` it will basically shutdown its own executor (which is why I stopped trying to steal SCMTrigger's queue because `setExecutor` shuts down the old one (without a no-op check beforehand... Nice eh))

So yes the admin will have 110 threads until the 100 previously executing tasks complete. No need to wait for GC

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java (hunk 3 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: Because we want to catch when the thread pool size is changed in any way (up or down)

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java (hunk 4 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: You mean `checkThreadPoolSizeAndUpdateIfNecessary` if you want to be exact. 

Most calls will just be a check so I named after the normal case

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java (hunk 2 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: I don't see the value, seems like change for change's sake. The other method is not too complex whereas we'd need to add a three way test to use this as an executor service factory method

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java (hunk 4 above)
Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: i'm for the verbose naming in such cases

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java (hunk 3 above)
Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: so we need the ability to change maximumThreads. Can't see it now

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java (hunk 2 above)
Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: Don't think the complexity reduction in such simple way is "like change for change's sake". At minimal state it should be named as `namedThreadFactory` as of its simple shortcut.

**Comment by jtnord** [comment]
Yes, my bad.  :bee:

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java (hunk 3 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: We piggy-back off `SCMTrigger.getPollingThreadCount()` which gives the UI. The `maximumThreads` field is a cache of the last seen value

Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: aha! a bug caught! cool!

**Comment by KostyaSha** [comment]
Btw, the same queue i'm using in other trigger plugins. Is it possible to make queue public and place configuration in global config?

**Comment by KostyaSha** [comment]
@reviewbybees @stephenc relogin!

**Comment by stephenc** [comment]
Nahh just depends which email I reply to!

**Comment by stephenc** [comment]
Anyway, my phone is the holder of the 2FA for that account...

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java [hunk 6]
```diff
@@ -235,6 +239,32 @@ public void writeLogTo(XMLOutput out) throws IOException {
...
+
+        public DescriptorImpl() {
+            checkThreadPoolSizeAndUpdateIfNecessary();
+        }
+
+        /**
+         * Update the {@link java.util.concurrent.ExecutorService} instance.
+         */
+        /*package*/
+        synchronized void checkThreadPoolSizeAndUpdateIfNecessary() {
```
Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: it should be a test for this logic

Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: I have lost any energy to continue this PR any further. I want to concentrate my time on the critical root cause, i.e. https://issues.jenkins-ci.org/browse/JENKINS-36147 which once implemented will mean we can rip out this code. If you want to add tests to this code which should be removed once I implement https://issues.jenkins-ci.org/browse/JENKINS-36147, feel free. I will not be updating this PR any further at this point.

Comment (KostyaSha) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: It would require updating core that is not good for this plugin. Or we can backport your core improvements after.

Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: well my plan is to make the API available via scm-api without a core bump... and then on newer cores it either will revert to the core implementation (classloader priority) or the scm-api variant will use reflection to redirect to the core implementation

Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: However, it will take me a while to get that API right (as I don't want to end up with the same mistakes of the main Queue API and end up having to have a "big fat lock") so for now, to save users from commit storms killing their Jenkins, this fix would be good to have released.

Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: ookay, i'll merge this. But this is an example of permanent state "will fix that better in future releases". It can't be finished while you've start do things right everywhere
